File: Ulysse/Modules/Module.m
//
//  Module.m
//  Ulysse
//

#import "Module.h"

NSMutableDictionary<NSString*, Module*> *modules = nil;

@implementation Module

+ (instancetype)moduleWithName:(NSString *)name {
  if (!modules) {
    modules = [NSMutableDictionary dictionary];
  }
  Module *module = [modules objectForKey:name];
  if (!module) {
    module = [[[self class] alloc] initWithName:name];
    [modules setObject:module forKey:name];
  }
  return module;
}

- (instancetype)initWithName:(NSString *)name {
  self = [super init];
  if (self) {
    _name = name;
  }
  return self;
}

@end

File: Ulysse/Modules/Module.swift
import Foundation

class Module: NSObject {

  @objc let name: String
  @objc private(set) var sortedKeys: Array<String> = [String]()
  @objc private(set) var values: Dictionary<String, Any> = [String: Any]()
  
  init(name: String) {
    self.name = name
    super.init()
  }

  func setAllValues(_ values: Dictionary<String, Any>) {
    self.values = values
    var newSortedKeys: Array = Array(self.values.keys)
    newSortedKeys = newSortedKeys.sorted{ $0.compare($1, options: .caseInsensitive) == .orderedAscending }
    self.sortedKeys = newSortedKeys.filter { $0 != "errors" }
    self.values.removeValue(forKey: "errors")
  }

  func value(key: String) -> Any? {
    return self.values[key]
  }

  func humanValue(key: String, short: Bool) -> Any? {
    return self.value(key: key)
  }

  func humanKey(key: String) -> String {
    return key
  }
}
